### ISSUE TITLE
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type

### ISSUE THREAD

**Issue by TheDutchCoder** [microsoft/TypeScript#39128]
**TypeScript Version:**  3.9.5

**Search Terms:**
Element implicitly has an 'any' type

**Code**

```ts
const addressComponent: google.maps.GeocoderAddressComponent = someResult
let foo = 'short_name'
console.log(addressComponent[foo])
console.log(addressComponent['short_name']) // This gives the Implicit any error
```

**Expected behavior:**
The error not to be thrown, both indices are strings, so there's no reason why a variable can't be used!?

**Actual behavior:**
Error thrown when using a variable with an implicit or explicit string type

**Comment by RyanCavanaugh** [comment]
Please provide a self-contained repro; the situation is entirely dependent on the declarations of `google.maps.GeocoderAddressComponent` and `someResult`

**Comment by TheDutchCoder** [comment]
It is just using the already existing type definitions: http://definitelytyped.org/docs/googlemaps--google.maps/interfaces/google.maps.geocoderaddresscomponent.html#long_name

The issue is that TS is "forcing" to be more specific about a type, even though the interface itself doesn't contain more specific types.

Now I practically have to create a mock interface, that just copies whatever is already in the type definition and use:

```
const foo: keyof SomeMockInterface = 'short_name'
```

**Comment by RyanCavanaugh** [comment]
Unfortunately without a self-contained repro we cannot advise further

**Comment by TheDutchCoder** [comment]
https://codesandbox.io/s/sleepy-maxwell-p1xf6?file=/src/index.ts

Here you go

**Comment by RyanCavanaugh** [comment]
What I mean by "self-contained" is that there aren't unresolved identifiers.

Anyway this is the expected behavior; object types are not sealed and indexing by an arbitrary `string` is not at all guaranteed to produce one of the declared property types. When TS has control flow and initialization information to see that the string *isn't* an arbitrary one, which I am guessing is what you were trying to show in the OP, then different behavior can occur.

This can be followed up on on Stack Overflow, etc, since this isn't at all indicative of a bug.

**Comment by TheDutchCoder** [comment]
Why would TS have no control flow over a variable?
That's the part that doesn't make sense to me.

```
const foo = {
  thing: 'thing'
}

const a = 'thing'
const b = foo[a]
const c = foo['thing']
```

Both are indexing by "an arbitrary string", both `a` and `'thing'` have the same type (`string`). Unless I'm missing something obvious?

**Comment by TheDutchCoder** [comment]
I still don't think this is "working as intended".
The error thrown is that "string" can't be used as the index type, which is just not true, since all indices are strings for that interface.

I think what TS really means is that only *certain* strings can be used as the index type (namely the `google.maps.GeocoderAddressComponent` indices).

If TS knows this, it should tell me, and not incorrectly state that strings can't be used, which is simply not correct. It should tell me that it expects a certain type of string: `google.maps.GeocoderAddressComponent`

**Comment by RyanCavanaugh** [comment]
@TheDutchCoder you haven't found a 5-year-old bug in the property lookup logic. TS can see that `"thing"` is the name of a property on the object; if you change it to `"thing2"` you'll see it behaves the same way as "an arbitrary string"

**Comment by klesun** [comment]
tl;dr
```ts
let foo = 'short_name' as const;
```
